File: csharp/src/Yosina/TransliteratorRecipe.cs
// Copyright (c) Yosina. All rights reserved.

using System.Runtime.InteropServices;
using Charset = Yosina.Transliterators.IvsSvsBaseTransliterator.Charset;
using Mapping = Yosina.Transliterators.HyphensTransliterator.Mapping;

namespace Yosina;
/// <summary>
/// Configuration recipe for building transliterator chains. This record provides a declarative way to
/// configure complex transliterator chains using high-level options that are automatically converted
/// to the appropriate transliterator configurations.
/// </summary>
public record class TransliteratorRecipe
{
    /// <summary>Options for full-width conversion.</summary>
    [StructLayout(LayoutKind.Auto)]
    public readonly struct ToFullwidthOptions
    {
        public static readonly ToFullwidthOptions Disabled = new(false, false);
        public static readonly ToFullwidthOptions Enabled = new(true, false);
        public static readonly ToFullwidthOptions U005cAsYenSign = new(true, true);

        private readonly bool enabled;
        private readonly bool u005cAsYenSign;

        public bool IsEnabled => this.enabled;

        public bool IsU005cAsYenSign => this.u005cAsYenSign;

        public static implicit operator ToFullwidthOptions(bool enabled)
        {
            return enabled ? Enabled : Disabled;
        }

        public ToFullwidthOptions(bool enabled, bool u005cAsYenSign)
        {
            this.enabled = enabled;
            this.u005cAsYenSign = u005cAsYenSign;
        }
    }

    /// <summary>Options for half-width conversion.</summary>
    [StructLayout(LayoutKind.Auto)]
    public readonly struct ToHalfwidthOptions
    {
        public static readonly ToHalfwidthOptions Disabled = new(false, false);
        public static readonly ToHalfwidthOptions Enabled = new(true, false);
        public static readonly ToHalfwidthOptions HankakuKana = new(true, true);

        private readonly bool enabled;
        private readonly bool hankakuKana;

        public bool IsEnabled => this.enabled;

        public bool IsHankakuKana => this.hankakuKana;

        public static implicit operator ToHalfwidthOptions(bool enabled)
        {
            return enabled ? Enabled : Disabled;
        }

        public ToHalfwidthOptions(bool enabled, bool hankakuKana)
        {
            this.enabled = enabled;
            this.hankakuKana = hankakuKana;
        }
    }

    /// <summary>Options for IVS/SVS removal.</summary>
    [StructLayout(LayoutKind.Auto)]
    public readonly struct RemoveIvsSvsOptions
    {
        public static readonly RemoveIvsSvsOptions Disabled = new(false, false);
        public static readonly RemoveIvsSvsOptions Enabled = new(true, false);
        public static readonly RemoveIvsSvsOptions DropAllSelectors = new(true, true);

        private readonly bool enabled;
        private readonly bool dropAllSelectors;

        public bool IsEnabled => this.enabled;

        public bool IsDropAllSelectors => this.dropAllSelectors;

        public static implicit operator RemoveIvsSvsOptions(bool enabled)
        {
            return enabled ? Enabled : Disabled;
        }

        private RemoveIvsSvsOptions(bool enabled, bool dropAllSelectors)
        {
            this.enabled = enabled;
            this.dropAllSelectors = dropAllSelectors;
        }
    }

    /// <summary>Options for hyphens replacement.</summary>
    public readonly struct ReplaceHyphensOptions
    {
        public static readonly ReplaceHyphensOptions Disabled = new(false, null);
        public static readonly ReplaceHyphensOptions Enabled = new(true, new List<Mapping> { Mapping.JISX0208_90_WINDOWS, Mapping.JISX0201 });

        private readonly bool enabled;
        private readonly IList<Mapping>? precedence;

        public bool IsEnabled => this.enabled;

        public IList<Mapping>? Precedence => this.precedence;

        public static ReplaceHyphensOptions WithPrecedence(IList<Mapping> precedence)
        {
            return new ReplaceHyphensOptions(true, precedence);
        }

        public static implicit operator ReplaceHyphensOptions(bool enabled)
        {
            return enabled ? Enabled : Disabled;
        }

        private ReplaceHyphensOptions(bool enabled, IList<Mapping>? precedence)
        {
            this.enabled = enabled;
            this.precedence = precedence;
        }
    }

    /// <summary>Options for circled or squared characters replacement.</summary>
    [StructLayout(LayoutKind.Auto)]
    public readonly struct ReplaceCircledOrSquaredCharactersOptions
    {
        public static readonly ReplaceCircledOrSquaredCharactersOptions Disabled = new(false, false);
        public static readonly ReplaceCircledOrSquaredCharactersOptions Enabled = new(true, true);
        public static readonly ReplaceCircledOrSquaredCharactersOptions ExcludeEmojis = new(true, false);

        private readonly bool enabled;
        private readonly bool includeEmojis;

        public bool IsEnabled => this.enabled;

        public bool IncludeEmojis => this.includeEmojis;

        public static implicit operator ReplaceCircledOrSquaredCharactersOptions(bool enabled)
        {
            return enabled ? Enabled : Disabled;
        }

        private ReplaceCircledOrSquaredCharactersOptions(bool enabled, bool includeEmojis)
        {
            this.enabled = enabled;
            this.includeEmojis = includeEmojis;
        }
    }

    // Recipe properties with default values

    /// <summary>
    /// Gets a value indicating whether replace codepoints that correspond to old-style kanji glyphs (旧字体; kyu-ji-tai) with their modern equivalents (新字体; shin-ji-tai).
    /// </summary>
    /// <example>
    /// <code>
    /// // Input:  "舊字體の變換"
    /// // Output: "旧字体の変換"
    /// var recipe = new TransliteratorRecipe { KanjiOldNew = true };
    /// </code>
    /// </example>
    public bool KanjiOldNew { get; init; }

    /// <summary>
    /// Gets a value indicating whether replace "suspicious" hyphens with prolonged sound marks, and vice versa.
    /// </summary>
    /// <example>
    /// <code>
    /// // Input:  "スーパ-" (with hyphen-minus)
    /// // Output: "スーパー" (becomes prolonged sound mark)
    /// var recipe = new TransliteratorRecipe { ReplaceSuspiciousHyphensToProlongedSoundMarks = true };
    /// </code>
    /// </example>
    public bool ReplaceSuspiciousHyphensToProlongedSoundMarks { get; init; }

    /// <summary>
    /// Gets a value indicating whether replace combined characters with their corresponding characters.
    /// </summary>
    /// <example>
    /// <code>
    /// // Input:  "㍻" (single character for Heisei era)
    /// // Output: "平成"
    /// // Input:  "㈱"
    /// // Output: "(株)"
    /// var recipe = new TransliteratorRecipe { ReplaceCombinedCharacters = true };
    /// </code>
    /// </example>
    public bool ReplaceCombinedCharacters { get; init; }

    /// <summary>
    /// Gets replace circled or squared characters with their corresponding templates.
    /// </summary>
    /// <example>
    /// <code>
    /// // Input:  "①②③"
    /// // Output: "(1)(2)(3)"
    /// // Input:  "㊙㊗"
    /// // Output: "(秘)(祝)"
    /// var recipe = new TransliteratorRecipe { ReplaceCircledOrSquaredCharacters = ReplaceCircledOrSquaredCharactersOptions.Enabled };
    /// </code>
    /// </example>
    public ReplaceCircledOrSquaredCharactersOptions ReplaceCircledOrSquaredCharacters { get; init; } = ReplaceCircledOrSquaredCharactersOptions.Disabled;

    /// <summary>
    /// Gets a value indicating whether replace ideographic annotations used in the traditional method of Chinese-to-Japanese translation devised in ancient Japan.
    /// </summary>
    /// <example>
    /// <code>
    /// // Input:  "㆖㆘" (ideographic annotations)
    /// // Output: "上下"
    /// var recipe = new TransliteratorRecipe { ReplaceIdeographicAnnotations = true };
    /// </code>
    /// </example>
    public bool ReplaceIdeographicAnnotations { get; init; }

    /// <summary>
    /// Gets a value indicating whether replace codepoints for the Kang Xi radicals whose glyphs resemble those of CJK ideographs with the CJK ideograph counterparts.
    /// </summary>
    /// <example>
    /// <code>
    /// // Input:  "⾔⾨⾷" (Kangxi radicals)
    /// // Output: "言門食" (CJK ideographs)
    /// var recipe = new TransliteratorRecipe { ReplaceRadicals = true };
    /// </code>
    /// </example>
    public bool ReplaceRadicals { get; init; }

    /// <summary>
    /// Gets a value indicating whether replace various space characters with plain whitespaces or empty strings.
    /// </summary>
    /// <example>
    /// <code>
    /// // Input:  "A　B" (ideographic space U+3000)
    /// // Output: "A B" (half-width space)
    /// // Input:  "A B" (non-breaking space U+00A0)
    /// // Output: "A B" (regular space)
    /// var recipe = new TransliteratorRecipe { ReplaceSpaces = true };
    /// </code>
    /// </example>
    public bool ReplaceSpaces { get; init; }

    /// <summary>
    /// Gets replace various dash or hyphen symbols with those common in Japanese writing.
    /// </summary>
    /// <example>
    /// <code>
    /// // Input:  "2019—2020" (em dash)
    /// // Output: "2019-2020" (hyphen-minus)
    /// // Input:  "A–B" (en dash)
    /// // Output: "A-B"
    /// var recipe = new TransliteratorRecipe { ReplaceHyphens = ReplaceHyphensOptions.Enabled };
    /// </code>
    /// </example>
    public ReplaceHyphensOptions ReplaceHyphens { get; init; } = ReplaceHyphensOptions.Disabled;

    /// <summary>
    /// Gets a value indicating whether replace mathematical alphanumerics with their plain ASCII equivalents.
    /// </summary>
    /// <example>
    /// <code>
    /// // Input:  "𝐀𝐁𝐂" (mathematical bold)
    /// // Output: "ABC"
    /// // Input:  "𝟏𝟐𝟑" (mathematical bold digits)
    /// // Output: "123"
    /// var recipe = new TransliteratorRecipe { ReplaceMathematicalAlphanumerics = true };
    /// </code>
    /// </example>
    public bool ReplaceMathematicalAlphanumerics { get; init; }

    /// <summary>
    /// Gets a value indicating whether combine decomposed hiraganas and katakanas into single counterparts.
    /// </summary>
    /// <example>
    /// <code>
    /// // Input:  "が" (か + ゙)
    /// // Output: "が" (single character)
    /// // Input:  "ヘ゜" (ヘ + ゜)
    /// // Output: "ペ" (single character)
    /// var recipe = new TransliteratorRecipe { CombineDecomposedHiraganasAndKatakanas = true };
    /// </code>
    /// </example>
    public bool CombineDecomposedHiraganasAndKatakanas { get; init; }

    /// <summary>
    /// Gets replace half-width characters to fullwidth equivalents.
    /// </summary>
    /// <example>
    /// <code>
    /// // Input:  "ABC123"
    /// // Output: "ＡＢＣ１２３"
    /// // Input:  "ｶﾀｶﾅ"
    /// // Output: "カタカナ"
    /// var recipe = new TransliteratorRecipe { ToFullwidth = ToFullwidthOptions.Enabled };
    /// </code>
    /// </example>
    public ToFullwidthOptions ToFullwidth { get; init; } = ToFullwidthOptions.Disabled;

    /// <summary>
    /// Gets replace full-width characters with their half-width equivalents.
    /// </summary>
    /// <example>
    /// <code>
    /// // Input:  "ＡＢＣ１２３"
    /// // Output: "ABC123"
    /// // Input:  "カタカナ" (with hankaku-kana)
    /// // Output: "ｶﾀｶﾅ"
    /// var recipe = new TransliteratorRecipe { ToHalfwidth = ToHalfwidthOptions.HankakuKana };
    /// </code>
    /// </example>
    public ToHalfwidthOptions ToHalfwidth { get; init; } = ToHalfwidthOptions.Disabled;

    /// <summary>
    /// Gets replace CJK ideographs followed by IVSes and SVSes with those without selectors based on Adobe-Japan1 character mappings.
    /// </summary>
    /// <example>
    /// <code>
    /// // Input:  "葛󠄀" (葛 + IVS U+E0100)
    /// // Output: "葛" (without selector)
    /// // Input:  "辻󠄀" (辻 + IVS)
    /// // Output: "辻"
    /// var recipe = new TransliteratorRecipe { RemoveIvsSvs = RemoveIvsSvsOptions.Enabled };
    /// </code>
    /// </example>
    public RemoveIvsSvsOptions RemoveIvsSvs { get; init; } = RemoveIvsSvsOptions.Disabled;

    public Charset Charset { get; init; } = Charset.UniJis2004;

    // Builder methods for fluent API
    public TransliteratorRecipe WithKanjiOldNew(bool kanjiOldNew) => this with { KanjiOldNew = kanjiOldNew };

    public TransliteratorRecipe WithReplaceSuspiciousHyphensToProlongedSoundMarks(bool replace) => this with { ReplaceSuspiciousHyphensToProlongedSoundMarks = replace };

    public TransliteratorRecipe WithReplaceCombinedCharacters(bool replace) => this with { ReplaceCombinedCharacters = replace };

    public TransliteratorRecipe WithReplaceCircledOrSquaredCharacters(ReplaceCircledOrSquaredCharactersOptions options) => this with { ReplaceCircledOrSquaredCharacters = options };

    public TransliteratorRecipe WithReplaceIdeographicAnnotations(bool replace) => this with { ReplaceIdeographicAnnotations = replace };

    public TransliteratorRecipe WithReplaceRadicals(bool replace) => this with { ReplaceRadicals = replace };

    public TransliteratorRecipe WithReplaceSpaces(bool replace) => this with { ReplaceSpaces = replace };

    public TransliteratorRecipe WithReplaceHyphens(ReplaceHyphensOptions options) => this with { ReplaceHyphens = options };

    public TransliteratorRecipe WithReplaceMathematicalAlphanumerics(bool replace) => this with { ReplaceMathematicalAlphanumerics = replace };

    public TransliteratorRecipe WithCombineDecomposedHiraganasAndKatakanas(bool combine) => this with { CombineDecomposedHiraganasAndKatakanas = combine };

    public TransliteratorRecipe WithToFullwidth(ToFullwidthOptions options) => this with { ToFullwidth = options };

    public TransliteratorRecipe WithToHalfwidth(ToHalfwidthOptions options) => this with { ToHalfwidth = options };

    public TransliteratorRecipe WithRemoveIvsSvs(RemoveIvsSvsOptions options) => this with { RemoveIvsSvs = options };

    public TransliteratorRecipe WithCharset(Charset charset) => this with { Charset = charset };

    /// <summary>
    /// Build transliterator configurations from this recipe.
    /// </summary>
    /// <returns>A list of TransliteratorConfig objects that can be passed to Yosina.MakeTransliterator.</returns>
    /// <exception cref="ArgumentException">If the recipe contains mutually exclusive options.</exception>
    public List<TransliteratorConfig> BuildTransliteratorConfigs()
    {
        // Check for mutually exclusive options
        var errors = new List<string>();
        if (this.ToFullwidth.IsEnabled && this.ToHalfwidth.IsEnabled)
        {
            errors.Add("toFullwidth and toHalfwidth are mutually exclusive");
        }

        if (errors.Any())
        {
            throw new ArgumentException(string.Join("; ", errors));
        }

        var ctx = new TransliteratorConfigListBuilder();

        // Apply transformations in the specified order
        ctx = this.ApplyKanjiOldNew(ctx);
        ctx = this.ApplyReplaceSuspiciousHyphensToProlongedSoundMarks(ctx);
        ctx = this.ApplyReplaceCircledOrSquaredCharacters(ctx);
        ctx = this.ApplyReplaceCombinedCharacters(ctx);
        ctx = this.ApplyReplaceIdeographicAnnotations(ctx);
        ctx = this.ApplyReplaceRadicals(ctx);
        ctx = this.ApplyReplaceSpaces(ctx);
        ctx = this.ApplyReplaceHyphens(ctx);
        ctx = this.ApplyReplaceMathematicalAlphanumerics(ctx);
        ctx = this.ApplyCombineDecomposedHiraganasAndKatakanas(ctx);
        ctx = this.ApplyToFullwidth(ctx);
        ctx = this.ApplyToHalfwidth(ctx);
        ctx = this.ApplyRemoveIvsSvs(ctx);

        return ctx.Build();
    }

    private TransliteratorConfigListBuilder RemoveIvsSvsHelper(TransliteratorConfigListBuilder ctx, bool dropAllSelectors)
    {
        // First insert IVS-or-SVS mode at head
        ctx = ctx.InsertHead(
            new TransliteratorConfig("ivs-svs-base", new { Mode = "IvsSvs", this.Charset }),
            true);
        ctx = ctx.InsertTail(
            new TransliteratorConfig("ivs-svs-base", new { Mode = "Base", DropSelectorsAltogether = dropAllSelectors, this.Charset }),
            true);

        return ctx;
    }

    private TransliteratorConfigListBuilder ApplyKanjiOldNew(TransliteratorConfigListBuilder ctx)
    {
        if (this.KanjiOldNew)
        {
            ctx = this.RemoveIvsSvsHelper(ctx, false);
            ctx = ctx.InsertMiddle(new TransliteratorConfig("kanji-old-new"), false);
        }

        return ctx;
    }

    private TransliteratorConfigListBuilder ApplyReplaceSuspiciousHyphensToProlongedSoundMarks(TransliteratorConfigListBuilder ctx)
    {
        if (this.ReplaceSuspiciousHyphensToProlongedSoundMarks)
        {
            ctx = ctx.InsertMiddle(
                new TransliteratorConfig("prolonged-sound-marks", new { ReplaceProlongedMarksFollowingAlnums = true }),
                false);
        }

        return ctx;
    }

    private TransliteratorConfigListBuilder ApplyReplaceIdeographicAnnotations(TransliteratorConfigListBuilder ctx)
    {
        if (this.ReplaceIdeographicAnnotations)
        {
            ctx = ctx.InsertMiddle(new TransliteratorConfig("ideographic-annotations"), false);
        }

        return ctx;
    }

    private TransliteratorConfigListBuilder ApplyReplaceRadicals(TransliteratorConfigListBuilder ctx)
    {
        if (this.ReplaceRadicals)
        {
            ctx = ctx.InsertMiddle(new TransliteratorConfig("radicals"), false);
        }

        return ctx;
    }

    private TransliteratorConfigListBuilder ApplyReplaceSpaces(TransliteratorConfigListBuilder ctx)
    {
        if (this.ReplaceSpaces)
        {
            ctx = ctx.InsertMiddle(new TransliteratorConfig("spaces"), false);
        }

        return ctx;
    }

    private TransliteratorConfigListBuilder ApplyReplaceHyphens(TransliteratorConfigListBuilder ctx)
    {
        if (this.ReplaceHyphens.IsEnabled)
        {
            var options = this.ReplaceHyphens.Precedence != null
                ? new { Precedence = this.ReplaceHyphens.Precedence }
                : null;
            ctx = ctx.InsertMiddle(new TransliteratorConfig("hyphens", options), false);
        }

        return ctx;
    }

    private TransliteratorConfigListBuilder ApplyReplaceMathematicalAlphanumerics(TransliteratorConfigListBuilder ctx)
    {
        if (this.ReplaceMathematicalAlphanumerics)
        {
            ctx = ctx.InsertMiddle(new TransliteratorConfig("mathematical-alphanumerics"), false);
        }

        return ctx;
    }

    private TransliteratorConfigListBuilder ApplyCombineDecomposedHiraganasAndKatakanas(TransliteratorConfigListBuilder ctx)
    {
        if (this.CombineDecomposedHiraganasAndKatakanas)
        {
            ctx = ctx.InsertHead(
                new TransliteratorConfig("hira-kata-composition", new { DecomposeFirst = true }),
                false);
        }

        return ctx;
    }

    private TransliteratorConfigListBuilder ApplyToFullwidth(TransliteratorConfigListBuilder ctx)
    {
        if (this.ToFullwidth.IsEnabled)
        {
            var options = new { FullwidthToHalfwidth = false, U005cAsYenSign = this.ToFullwidth.IsU005cAsYenSign };
            ctx = ctx.InsertTail(new TransliteratorConfig("jisx0201-and-alike", options), false);
        }

        return ctx;
    }

    private TransliteratorConfigListBuilder ApplyToHalfwidth(TransliteratorConfigListBuilder ctx)
    {
        if (this.ToHalfwidth.IsEnabled)
        {
            var options = new { FullwidthToHalfwidth = true, ConvertGL = true, ConvertGR = this.ToHalfwidth.IsHankakuKana };
            ctx = ctx.InsertTail(new TransliteratorConfig("jisx0201-and-alike", options), false);
        }

        return ctx;
    }

    private TransliteratorConfigListBuilder ApplyRemoveIvsSvs(TransliteratorConfigListBuilder ctx)
    {
        if (this.RemoveIvsSvs.IsEnabled)
        {
            ctx = this.RemoveIvsSvsHelper(ctx, this.RemoveIvsSvs.IsDropAllSelectors);
        }

        return ctx;
    }

    private TransliteratorConfigListBuilder ApplyReplaceCombinedCharacters(TransliteratorConfigListBuilder ctx)
    {
        if (this.ReplaceCombinedCharacters)
        {
            ctx = ctx.InsertMiddle(new TransliteratorConfig("combined"), false);
        }

        return ctx;
    }

    private TransliteratorConfigListBuilder ApplyReplaceCircledOrSquaredCharacters(TransliteratorConfigListBuilder ctx)
    {
        if (this.ReplaceCircledOrSquaredCharacters.IsEnabled)
        {
            var options = new { includeEmojis = this.ReplaceCircledOrSquaredCharacters.IncludeEmojis };
            ctx = ctx.InsertMiddle(new TransliteratorConfig("circled-or-squared", options), false);
        }

        return ctx;
    }
}

File: dart/lib/src/transliterator_recipe.dart
import 'chained_transliterator.dart';
import 'transliterator.dart';
import 'transliterator_registry.dart';

// MARK: - Option Classes

/// Options for full-width conversion.
class ToFullwidthOptions {
  const ToFullwidthOptions({
    required this.enabled,
    required this.u005cAsYenSign,
  });

  const ToFullwidthOptions.disabled()
      : enabled = false,
        u005cAsYenSign = false;

  const ToFullwidthOptions.enabled()
      : enabled = true,
        u005cAsYenSign = false;

  const ToFullwidthOptions.u005cAsYenSign()
      : enabled = true,
        u005cAsYenSign = true;

  final bool enabled;
  final bool u005cAsYenSign;

  bool get isEnabled => enabled;
  bool get isU005cAsYenSign => u005cAsYenSign;

  static ToFullwidthOptions fromBool(bool enabled) {
    return enabled
        ? const ToFullwidthOptions.enabled()
        : const ToFullwidthOptions.disabled();
  }
}

/// Options for half-width conversion.
class ToHalfwidthOptions {
  const ToHalfwidthOptions({
    required this.enabled,
    required this.hankakuKana,
  });

  const ToHalfwidthOptions.disabled()
      : enabled = false,
        hankakuKana = false;

  const ToHalfwidthOptions.enabled()
      : enabled = true,
        hankakuKana = false;

  const ToHalfwidthOptions.hankakuKana()
      : enabled = true,
        hankakuKana = true;

  final bool enabled;
  final bool hankakuKana;

  bool get isEnabled => enabled;
  bool get isHankakuKana => hankakuKana;

  static ToHalfwidthOptions fromBool(bool enabled) {
    return enabled
        ? const ToHalfwidthOptions.enabled()
        : const ToHalfwidthOptions.disabled();
  }
}

/// Options for IVS/SVS removal.
class RemoveIvsSvsOptions {
  const RemoveIvsSvsOptions({
    required this.enabled,
    required this.dropAllSelectors,
  });

  const RemoveIvsSvsOptions.disabled()
      : enabled = false,
        dropAllSelectors = false;

  const RemoveIvsSvsOptions.enabled()
      : enabled = true,
        dropAllSelectors = false;

  const RemoveIvsSvsOptions.dropAllSelectors()
      : enabled = true,
        dropAllSelectors = true;

  final bool enabled;
  final bool dropAllSelectors;

  bool get isEnabled => enabled;
  bool get isDropAllSelectors => dropAllSelectors;

  static RemoveIvsSvsOptions fromBool(bool enabled) {
    return enabled
        ? const RemoveIvsSvsOptions.enabled()
        : const RemoveIvsSvsOptions.disabled();
  }
}

/// Options for hyphens replacement.
class ReplaceHyphensOptions {
  const ReplaceHyphensOptions({
    required this.enabled,
    this.precedence,
  });

  const ReplaceHyphensOptions.disabled()
      : enabled = false,
        precedence = null;

  const ReplaceHyphensOptions.enabled()
      : enabled = true,
        precedence = const ['jisx0208_90_windows', 'jisx0201'];

  final bool enabled;
  final List<String>? precedence;

  bool get isEnabled => enabled;

  static ReplaceHyphensOptions withPrecedence(List<String> precedence) {
    return ReplaceHyphensOptions(enabled: true, precedence: precedence);
  }

  static ReplaceHyphensOptions fromBool(bool enabled) {
    return enabled
        ? const ReplaceHyphensOptions.enabled()
        : const ReplaceHyphensOptions.disabled();
  }
}

/// Options for circled or squared characters replacement.
class ReplaceCircledOrSquaredCharactersOptions {
  const ReplaceCircledOrSquaredCharactersOptions({
    required this.enabled,
    required this.includeEmojis,
  });

  const ReplaceCircledOrSquaredCharactersOptions.disabled()
      : enabled = false,
        includeEmojis = false;

  const ReplaceCircledOrSquaredCharactersOptions.enabled()
      : enabled = true,
        includeEmojis = true;

  const ReplaceCircledOrSquaredCharactersOptions.excludeEmojis()
      : enabled = true,
        includeEmojis = false;

  final bool enabled;
  final bool includeEmojis;

  bool get isEnabled => enabled;

  static ReplaceCircledOrSquaredCharactersOptions fromBool(bool enabled) {
    return enabled
        ? const ReplaceCircledOrSquaredCharactersOptions.enabled()
        : const ReplaceCircledOrSquaredCharactersOptions.disabled();
  }
}

// MARK: - TransliteratorConfig

/// Configuration for a single transliterator.
class TransliteratorConfig {
  const TransliteratorConfig(this.name, [this.options = const {}]);

  factory TransliteratorConfig.fromMap(Map<String, dynamic> map) {
    return TransliteratorConfig(
      map['name'] as String,
      (map['options'] as Map<String, dynamic>?) ?? const {},
    );
  }

  final String name;
  final Map<String, dynamic> options;
}

// MARK: - TransliteratorRecipe

/// Represents a recipe for creating transliterators.
///
/// This class provides both low-level (direct config list) and high-level
/// (declarative options) ways to configure transliterator chains.
class TransliteratorRecipe {
  const TransliteratorRecipe(this.configs);

  /// Create a recipe from a list of transliterator names.
  factory TransliteratorRecipe.fromNames(List<String> names) {
    return TransliteratorRecipe(
      names.map((name) => TransliteratorConfig(name)).toList(),
    );
  }

  /// Create a recipe from a map of transliterator configurations.
  factory TransliteratorRecipe.fromMap(List<Map<String, dynamic>> configs) {
    return TransliteratorRecipe(
      configs.map((config) => TransliteratorConfig.fromMap(config)).toList(),
    );
  }

  /// Create a recipe using high-level configuration options.
  ///
  /// This factory provides a declarative way to configure complex transliterator
  /// chains using semantic options that are automatically converted to the
  /// appropriate transliterator configurations.
  factory TransliteratorRecipe.withOptions({
    bool kanjiOldNew = false,
    bool replaceSuspiciousHyphensToProlongedSoundMarks = false,
    bool replaceCombinedCharacters = false,
    ReplaceCircledOrSquaredCharactersOptions replaceCircledOrSquaredCharacters =
        const ReplaceCircledOrSquaredCharactersOptions.disabled(),
    bool replaceIdeographicAnnotations = false,
    bool replaceRadicals = false,
    bool replaceSpaces = false,
    ReplaceHyphensOptions replaceHyphens =
        const ReplaceHyphensOptions.disabled(),
    bool replaceMathematicalAlphanumerics = false,
    bool combineDecomposedHiraganasAndKatakanas = false,
    ToFullwidthOptions toFullwidth = const ToFullwidthOptions.disabled(),
    ToHalfwidthOptions toHalfwidth = const ToHalfwidthOptions.disabled(),
    RemoveIvsSvsOptions removeIvsSvs = const RemoveIvsSvsOptions.disabled(),
    String charset = 'unijis_2004',
  }) {
    // Check for mutually exclusive options
    final errors = <String>[];
    if (toFullwidth.isEnabled && toHalfwidth.isEnabled) {
      errors.add('toFullwidth and toHalfwidth are mutually exclusive');
    }

    if (errors.isNotEmpty) {
      throw ArgumentError(errors.join('; '));
    }

    final builder = _TransliteratorConfigListBuilder();

    // Apply transformations in the specified order
    _applyKanjiOldNew(builder, kanjiOldNew, removeIvsSvs, charset);
    _applyReplaceSuspiciousHyphensToProlongedSoundMarks(
        builder, replaceSuspiciousHyphensToProlongedSoundMarks);
    _applyReplaceCircledOrSquaredCharacters(
        builder, replaceCircledOrSquaredCharacters);
    _applyReplaceCombinedCharacters(builder, replaceCombinedCharacters);
    _applyReplaceIdeographicAnnotations(builder, replaceIdeographicAnnotations);
    _applyReplaceRadicals(builder, replaceRadicals);
    _applyReplaceSpaces(builder, replaceSpaces);
    _applyReplaceHyphens(builder, replaceHyphens);
    _applyReplaceMathematicalAlphanumerics(
        builder, replaceMathematicalAlphanumerics);
    _applyCombineDecomposedHiraganasAndKatakanas(
        builder, combineDecomposedHiraganasAndKatakanas);
    _applyToFullwidth(builder, toFullwidth);
    _applyToHalfwidth(builder, toHalfwidth);
    _applyRemoveIvsSvs(builder, removeIvsSvs, charset);

    return TransliteratorRecipe(builder.build());
  }

  final List<TransliteratorConfig> configs;

  /// Create a transliterator from this recipe using the given registry.
  Transliterator create(TransliteratorRegistry registry) {
    if (configs.isEmpty) {
      throw ArgumentError(
          'Recipe must contain at least one transliterator config');
    }

    if (configs.length == 1) {
      return registry.create(configs[0].name, configs[0].options);
    }

    final transliterators = configs
        .map((config) => registry.create(config.name, config.options))
        .toList();

    return ChainedTransliterator(transliterators);
  }

  // Private helper methods for applying transformations

  static void _removeIvsSvsHelper(_TransliteratorConfigListBuilder ctx,
      bool dropAllSelectors, String charset) {
    // First insert IVS-or-SVS mode at head
    ctx.insertHead(
      TransliteratorConfig('ivsSvsBase', {
        'mode': 'ivs-or-svs',
        'charset': charset,
      }),
      forceReplace: true,
    );
    ctx.insertTail(
      TransliteratorConfig('ivsSvsBase', {
        'mode': 'base',
        'dropSelectorsAltogether': dropAllSelectors,
        'charset': charset,
      }),
      forceReplace: true,
    );
  }

  static void _applyKanjiOldNew(_TransliteratorConfigListBuilder ctx,
      bool kanjiOldNew, RemoveIvsSvsOptions removeIvsSvs, String charset) {
    if (kanjiOldNew) {
      if (!removeIvsSvs.isEnabled) {
        _removeIvsSvsHelper(ctx, false, charset);
      }
      ctx.insertMiddle(
        const TransliteratorConfig('kanjiOldNew'),
        forceReplace: false,
      );
    }
  }

  static void _applyReplaceSuspiciousHyphensToProlongedSoundMarks(
      _TransliteratorConfigListBuilder ctx, bool replace) {
    if (replace) {
      ctx.insertMiddle(
        const TransliteratorConfig('prolongedSoundMarks', {
          'replaceProlongedMarksFollowingAlnums': true,
        }),
        forceReplace: false,
      );
    }
  }

  static void _applyReplaceIdeographicAnnotations(
      _TransliteratorConfigListBuilder ctx, bool replace) {
    if (replace) {
      ctx.insertMiddle(
        const TransliteratorConfig('ideographicAnnotations'),
        forceReplace: false,
      );
    }
  }

  static void _applyReplaceRadicals(
      _TransliteratorConfigListBuilder ctx, bool replace) {
    if (replace) {
      ctx.insertMiddle(
        const TransliteratorConfig('radicals'),
        forceReplace: false,
      );
    }
  }

  static void _applyReplaceSpaces(
      _TransliteratorConfigListBuilder ctx, bool replace) {
    if (replace) {
      ctx.insertMiddle(
        const TransliteratorConfig('spaces'),
        forceReplace: false,
      );
    }
  }

  static void _applyReplaceHyphens(
      _TransliteratorConfigListBuilder ctx, ReplaceHyphensOptions options) {
    if (options.isEnabled) {
      final Map<String, dynamic> configOptions =
          options.precedence != null ? {'precedence': options.precedence} : {};
      ctx.insertMiddle(
        TransliteratorConfig('hyphens', configOptions),
        forceReplace: false,
      );
    }
  }

  static void _applyReplaceMathematicalAlphanumerics(
      _TransliteratorConfigListBuilder ctx, bool replace) {
    if (replace) {
      ctx.insertMiddle(
        const TransliteratorConfig('mathematicalAlphanumerics'),
        forceReplace: false,
      );
    }
  }

  static void _applyCombineDecomposedHiraganasAndKatakanas(
      _TransliteratorConfigListBuilder ctx, bool combine) {
    if (combine) {
      ctx.insertHead(
        const TransliteratorConfig('hiraKataComposition', {
          'decomposeFirst': true,
        }),
        forceReplace: false,
      );
    }
  }

  static void _applyToFullwidth(
      _TransliteratorConfigListBuilder ctx, ToFullwidthOptions options) {
    if (options.isEnabled) {
      ctx.insertTail(
        TransliteratorConfig('jisX0201AndAlike', {
          'fullwidthToHalfwidth': false,
          'u005cAsYenSign': options.isU005cAsYenSign,
        }),
        forceReplace: false,
      );
    }
  }

  static void _applyToHalfwidth(
      _TransliteratorConfigListBuilder ctx, ToHalfwidthOptions options) {
    if (options.isEnabled) {
      ctx.insertTail(
        TransliteratorConfig('jisX0201AndAlike', {
          'fullwidthToHalfwidth': true,
          'convertGL': true,
          'convertGR': options.isHankakuKana,
        }),
        forceReplace: false,
      );
    }
  }

  static void _applyRemoveIvsSvs(_TransliteratorConfigListBuilder ctx,
      RemoveIvsSvsOptions options, String charset) {
    if (options.isEnabled) {
      _removeIvsSvsHelper(ctx, options.isDropAllSelectors, charset);
    }
  }

  static void _applyReplaceCombinedCharacters(
      _TransliteratorConfigListBuilder ctx, bool replace) {
    if (replace) {
      ctx.insertMiddle(
        const TransliteratorConfig('combined'),
        forceReplace: false,
      );
    }
  }

  static void _applyReplaceCircledOrSquaredCharacters(
      _TransliteratorConfigListBuilder ctx,
      ReplaceCircledOrSquaredCharactersOptions options) {
    if (options.isEnabled) {
      ctx.insertMiddle(
        TransliteratorConfig('circledOrSquared', {
          'includeEmojis': options.includeEmojis,
        }),
        forceReplace: false,
      );
    }
  }
}

// MARK: - Internal Builder

/// Internal builder for creating lists of transliterator configurations.
class _TransliteratorConfigListBuilder {
  final List<TransliteratorConfig> _head = [];
  final List<TransliteratorConfig> _tail = [];

  void insertHead(TransliteratorConfig config, {required bool forceReplace}) {
    final existingIndex = _findConfigIndex(_head, config.name);

    if (existingIndex >= 0) {
      if (forceReplace) {
        _head[existingIndex] = config;
      }
    } else {
      _head.insert(0, config);
    }
  }

  void insertMiddle(TransliteratorConfig config, {required bool forceReplace}) {
    final existingIndex = _findConfigIndex(_tail, config.name);

    if (existingIndex >= 0) {
      if (forceReplace) {
        _tail[existingIndex] = config;
      }
    } else {
      _tail.insert(0, config); // Insert at beginning of tail (middle position)
    }
  }

  void insertTail(TransliteratorConfig config, {required bool forceReplace}) {
    final existingIndex = _findConfigIndex(_tail, config.name);

    if (existingIndex >= 0) {
      if (forceReplace) {
        _tail[existingIndex] = config;
      }
    } else {
      _tail.add(config);
    }
  }

  int _findConfigIndex(List<TransliteratorConfig> configs, String name) {
    for (var i = 0; i < configs.length; i++) {
      if (configs[i].name == name) {
        return i;
      }
    }
    return -1;
  }

  List<TransliteratorConfig> build() {
    return [..._head, ..._tail];
  }
}
